tools/generate-examples.rb

```diff
@@ -0,0 +1,428 @@
+    # rubocop:disable Layout/LineLength
+    "https://github.com/oracle/graaljs/releases/download/graal-#{truffle_version}/graalnodejs#{community_edition ? "-community" : ""}-jvm-#{truffle_version}-macos-aarch64.tar.gz"
+    # rubocop:enable Layout/LineLength
+  end
+
+  def ruby_download_url
+    # rubocop:disable Layout/LineLength
+    "https://github.com/oracle/truffleruby/releases/download/graal-#{truffle_version}/truffleruby#{community_edition ? "-community" : ""}-jvm-#{truffle_version}-macos-aarch64.tar.gz"
+    # rubocop:enable Layout/LineLength
+  end
+end
+
+GraalVM_CE_23_1_2 = GraalVM.new("21.0.2", "23.1.2", true, "graalvm-ce-java21-23.1.2")
+GraalVM_GFTC_23_1_2 = GraalVM.new("21", "23.1.2", false, "graalvm-gftc-java21-23.1.2")
+
+GraalVM_CE_22_3_1 = GraalVMOld.new(
+  "https://github.com/graalvm/graalvm-ce-builds/releases/download/vm-22.3.1/graalvm-ce-java17-darwin-amd64-22.3.1.tar.gz",
+  "graalvm-ce-java17-22.3.1",
+)
+
+GraalVM_CE_21_2_0 = GraalVMOld.new(
+  "https://github.com/graalvm/graalvm-ce-builds/releases/download/vm-21.2.0/graalvm-ce-java11-darwin-amd64-21.2.0.tar.gz",
+  "graalvm-ce-java11-21.2.0",
+)
+
+GRAAL_VMS = [
+  GraalVM_CE_23_1_2,
+  GraalVM_GFTC_23_1_2,
+  GraalVM_CE_22_3_1,
+  GraalVM_CE_21_2_0,
+]
+
+REFERENCE_GRAALVM = GraalVM_GFTC_23_1_2
+
+def reference_graalvm?(graalvm)
+  graalvm == REFERENCE_GRAALVM
+end
+
+def run_command(graalvm, args, options = {})
+  if VERBOSE
+    log(graalvm, "  Command: #{args.join(" ")}")
+  end
+
+  IO.popen(args, options)
+end
+
+def run_java(graalvm, bin, dump_level, *args)
+  FileUtils.rm_rf("./graal_dumps")
+
+  pipe = run_command(graalvm, [
+    "#{bin}/java", \
+    "-XX:+PrintCompilation",
+    "-Dgraal.PrintGraphWithSchedule=true",
+    "-Dgraal.Dump=:#{dump_level}",
+  ] + args)
+
+  yield pipe
+
+  sleep(3)
+  Process.kill("KILL", pipe.pid)
+  Process.wait(pipe.pid)
+end
+
+def run_js(graalvm, bin, *args)
+  FileUtils.rm_rf("graal_dumps")
+
+  pipe = run_command(
+    graalvm,
+    [
+      "#{bin}/node",
+      "--experimental-options",
+      "--engine.CompileOnly=fib",
+      "--engine.TraceCompilation",
+      "--engine.MultiTier=false",
+      graalvm.options_version == 1 ? "--engine.Inlining=false" : "--compiler.Inlining=false",
+      "--vm.Dgraal.Dump=Truffle:1",
+    ] + args,
+    { err: [:child, :out] },
+  )
+
+  yield pipe
+
+  sleep(3)
+  Process.kill("KILL", pipe.pid)
+  Process.wait(pipe.pid)
+end
+
+def run_ruby(graalvm, bin, *args)
+  FileUtils.rm_rf("./graal_dumps")
+
+  pipe = run_command(
+    graalvm,
+    [
+      "#{bin}/ruby",
+      "--jvm",
+      "--experimental-options",
+      "--engine.TraceCompilation",
+      graalvm.options_version == 1 ? "--engine.NodeSourcePositions" : "--compiler.NodeSourcePositions",
+      "--engine.MultiTier=false",
+      "--vm.Dgraal.Dump=Truffle:1",
+    ] + args,
+    { err: [:child, :out] },
+  )
+
+  yield pipe
+
+  sleep(3)
+  Process.kill("KILL", pipe.pid)
+  Process.wait(pipe.pid)
+end
```

Comment:
There's quite a bit more functionality duplicated across the methods that you could move to this function.
```suggestion
def run_command(graalvm, args, options = {})
  if VERBOSE
    log(graalvm, "  Command: #{args.join(" ")}")
  end

  FileUtils.rm_rf("./graal_dumps")

  pipe = IO.popen(args, options)

  yield pipe

  sleep(3)
  Process.kill("KILL", pipe.pid)
  Process.wait(pipe.pid)
end

def run_java(graalvm, bin, dump_level, *args)
  run_command(graalvm, [
    "#{bin}/java", \
    "-XX:+PrintCompilation",
    "-Dgraal.PrintGraphWithSchedule=true",
    "-Dgraal.Dump=:#{dump_level}",
  ] + args)
end

def run_js(graalvm, bin, *args)
  run_command(
    graalvm,
    [
      "#{bin}/node",
      "--experimental-options",
      "--engine.CompileOnly=fib",
      "--engine.TraceCompilation",
      "--engine.MultiTier=false",
      graalvm.options_version == 1 ? "--engine.Inlining=false" : "--compiler.Inlining=false",
      "--vm.Dgraal.Dump=Truffle:1",
    ] + args,
    { err: [:child, :out] },
  )
end

def run_ruby(graalvm, bin, *args)
  run_command(
    graalvm,
    [
      "#{bin}/ruby",
      "--jvm",
      "--experimental-options",
      "--engine.TraceCompilation",
      graalvm.options_version == 1 ? "--engine.NodeSourcePositions" : "--compiler.NodeSourcePositions",
      "--engine.MultiTier=false",
      "--vm.Dgraal.Dump=Truffle:1",
    ] + args,
    { err: [:child, :out] },
  )
en
```

Reply:
I'm okay with the duplication. All I wanted was a way to log the command easily. I'm already a bit uneasy with the multiple levels required to execute a command. I think the simplicity of the original bash-like implementation was a lot easier to follow.